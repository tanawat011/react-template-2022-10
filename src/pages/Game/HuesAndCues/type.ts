/* istanbul ignore file */
import type { DocumentData, QueryDocumentSnapshot } from 'firebase/firestore'

export type Player = {
  id: string
  name: string
}

export type Room = {
  id: string
  password: string
  hintChoice: string[]
  hintSelected: string
  isStarted: boolean
  isSubmitResult: boolean
  totalRound: number
}

export type RawRoomPlayer = {
  refPlayer: QueryDocumentSnapshot<DocumentData>
  color: string
  score: number
  allSelected: string[]
  isOwner: boolean
  isTurn: boolean
  isHinter: boolean
  totalTurn: number
  seq: number
}

export type RoomPlayer = Omit<RawRoomPlayer, 'refPlayer'> & {
  player: Player
}
