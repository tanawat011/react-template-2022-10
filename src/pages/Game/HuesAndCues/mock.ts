/* istanbul ignore file */
import type { Player, Room, RoomPlayer } from './type'

type OptionAllPlayers = {
  p1?: Partial<RoomPlayer>
  p2?: Partial<RoomPlayer>
  p3?: Partial<RoomPlayer>
}

export const getMockHintChoice = ['A1', 'A2', 'A3', 'A4']

export const getMockRoom = (option?: Partial<Room>): Room => {
  return {
    id: 'room-id-x',
    password: '',
    hintChoice: [],
    hintSelected: '',
    isStarted: false,
    isSubmitResult: false,
    totalRound: 0,
    ...option,
  }
}

export const getMockPlayers = (): Player[] => {
  return [
    {
      id: 'player-id-1',
      name: 'p1',
    },
    {
      id: 'player-id-2',
      name: 'p2',
    },
    {
      id: 'player-id-3',
      name: 'p3',
    },
  ]
}

const defaultPlayerData = {
  score: 0,
  allSelected: [],
  isHinter: false,
  isOwner: false,
  isTurn: false,
  totalTurn: 0,
}

export const getMockPlayersInRoom = (option?: OptionAllPlayers): RoomPlayer[] => {
  const players = getMockPlayers()

  return [
    {
      player: players[0],
      color: 'bg-[#000000]',
      seq: 1,
      ...defaultPlayerData,
      ...option?.p1,
    },
    {
      player: players[1],
      color: 'bg-[#999999]',
      seq: 2,
      ...defaultPlayerData,
      ...option?.p2,
    },
    {
      player: players[2],
      color: 'bg-[#666666]',
      seq: 3,
      ...defaultPlayerData,
      ...option?.p3,
    },
  ]
}
