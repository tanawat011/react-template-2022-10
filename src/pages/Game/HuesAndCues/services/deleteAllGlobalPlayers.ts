import type { RoomPlayer } from '../type'

import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteDocument } from 'helpers/firebase'

export const deleteAllGlobalPlayers = async (players: RoomPlayer[]) => {
  const playerPath = FIRESTORE_PATH.DOC_PLAYER

  await Promise.all(
    players.map(async ({ player: { id } }) => {
      const path = playerPath + `/${id}`

      await deleteDocument(path)
    }),
  )
}
