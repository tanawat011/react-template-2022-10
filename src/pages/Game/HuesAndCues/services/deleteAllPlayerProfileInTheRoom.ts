import type { RoomPlayer } from '../_type'

import { getFirestore } from 'firebase/firestore'

import { deleteDocument } from 'helpers/firebase'

import { FIRESTORE_PATH } from '../_constants'

export const deleteAllPlayerProfileInTheRoom = async (players: RoomPlayer[]) => {
  const db = getFirestore()

  await Promise.all(
    players.map(async ({ uid }) => {
      const collectionPath = `${FIRESTORE_PATH.PLAYERS_COLLECTION}/${uid}`

      await deleteDocument(db, collectionPath)
    }),
  )
}
