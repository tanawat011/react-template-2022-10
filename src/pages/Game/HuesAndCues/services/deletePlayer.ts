import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deletePlayer = async (playerId: string) => {
  const path = FIRESTORE_PATH.DOC_PLAYER + `/${playerId}`

  await deleteCollection(path)
}
