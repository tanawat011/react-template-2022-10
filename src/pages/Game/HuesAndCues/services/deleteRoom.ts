import { FIRESTORE_PATH } from 'constants/gameHuesAndCues'
import { deleteCollection } from 'helpers/firebase'

export const deleteRoom = async (roomId: string) => {
  const path = FIRESTORE_PATH.DOC_ROOM + `/${roomId}`

  await deleteCollection(path)
}
