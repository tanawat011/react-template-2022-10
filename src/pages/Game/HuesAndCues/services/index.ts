export * from './deleteAllPlayerProfileInTheRoom'
export * from './deleteRoom'
export * from './getAllPlayersInTheRoom'
export * from './getPlayerInTheRoom'
export * from './getPlayerProfile'
export * from './setPlayerProfile'
export * from './setPlayerInTheRoom'
export * from './subscribeRoom'

export * from './navigate'
