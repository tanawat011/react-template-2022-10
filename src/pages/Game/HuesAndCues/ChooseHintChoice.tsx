import type { Room, RoomPlayer } from './type'

import { Button } from 'components/Button'
import { IconCircleDense } from 'components/Icons'
import { huesAndCues } from 'data/HuesAndCues'

import { setRoom } from './services'

type Prop = {
  room: Room
  currRoomPlayer: RoomPlayer
}

export const ChooseHintChoice: React.FC<Prop> = ({ room, currRoomPlayer }) => {
  const handleSelectHint = async (colorId: string) => {
    await setRoom({ ...room, hintSelected: colorId })
  }

  const isHinter = currRoomPlayer.isHinter
  const isSelectedHint = Boolean(room.hintSelected)

  const isDisabled = [!isHinter, isSelectedHint].includes(true)

  return (
    <>
      {room.hintChoice.map((colorId) => {
        const data = huesAndCues.flat().find((c) => c.id === colorId)
        const isHintSelected = room.hintSelected === colorId

        return (
          <Button
            key={`choose-hint-choice-${data?.color}`}
            className={data?.color}
            disabled={isDisabled}
            onClick={() => handleSelectHint(colorId)}
          >
            {colorId}
            {isHintSelected && <IconCircleDense />}
          </Button>
        )
      })}
    </>
  )
}
