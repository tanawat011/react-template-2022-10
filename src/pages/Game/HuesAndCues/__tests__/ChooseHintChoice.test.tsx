import type { RenderResult } from '@testing-library/react'

import { fireEvent, screen, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

import { renderWithProviders } from 'helpers/test'

import { ChooseHintChoice } from '../ChooseHintChoice'
import { getMockHintChoice, getMockPlayersInRoom, getMockRoom } from '../mock'
import { setRoom } from '../services'

jest.mock('../services')

describe('<ChooseHintChoice />', () => {
  test('clickable', async () => {
    let el: RenderResult
    const room = getMockRoom({ hintChoice: getMockHintChoice })
    const players = getMockPlayersInRoom({ p1: { isHinter: true } })

    await act(async () => {
      el = renderWithProviders(<ChooseHintChoice room={room} currRoomPlayer={players[0]} />)
    })

    const button = screen.getAllByTestId('button')[0]
    fireEvent.click(button)

    await waitFor(() => {
      expect(setRoom).toHaveBeenCalledTimes(1)
      expect(el.asFragment()).toMatchSnapshot()
    })
  })

  test('disabled, when current player are not a `hinter`', async () => {
    const room = getMockRoom({ hintChoice: getMockHintChoice })
    const players = getMockPlayersInRoom({
      p1: { isHinter: true },
    })

    renderWithProviders(<ChooseHintChoice room={room} currRoomPlayer={players[1]} />)

    const button = screen.getAllByTestId('button')[0]

    expect(button).toHaveAttribute('disabled')
  })

  test('disabled, when `hinter` none selected hint', async () => {
    const room = getMockRoom({ hintChoice: getMockHintChoice })
    const players = getMockPlayersInRoom({
      p1: { isHinter: true },
    })

    renderWithProviders(<ChooseHintChoice room={room} currRoomPlayer={players[0]} />)

    const button = screen.getAllByTestId('button')[0]

    expect(button).not.toHaveAttribute('disabled')
  })

  test('disabled, when `hinter` selected hint', async () => {
    const room = getMockRoom({ hintChoice: getMockHintChoice, hintSelected: 'A1' })
    const players = getMockPlayersInRoom({
      p1: { isHinter: true },
    })

    renderWithProviders(<ChooseHintChoice room={room} currRoomPlayer={players[0]} />)

    const button = screen.getAllByTestId('button')[0]

    expect(button).toHaveAttribute('disabled')
  })
})
