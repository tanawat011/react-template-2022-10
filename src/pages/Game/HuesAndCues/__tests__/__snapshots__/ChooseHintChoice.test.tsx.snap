// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChooseHintChoice /> clickable 1`] = `
<DocumentFragment>
  <button
    class="bg-[#612b10]"
    data-testid="button"
  >
    A1
  </button>
  <button
    class="bg-[#6b2810]"
    data-testid="button"
  >
    A2
  </button>
  <button
    class="bg-[#752316]"
    data-testid="button"
  >
    A3
  </button>
  <button
    class="bg-[#861e1a]"
    data-testid="button"
  >
    A4
  </button>
</DocumentFragment>
`;
